src/core/QCheck2.mli

```diff
@@ -0,0 +1,1992 @@
+
+  val small_string : ?gen:char t -> string t
+  (** Builds a string generator, length is {!small_nat}.
+      Accepts an optional character generator (the default is {!char}).
+
+      Shrinks on the number of characters first, then on the characters.
+  *)
+
+  val pure : 'a -> 'a t
+  (** [pure a] creates a generator that always returns [a].
+
+      Does not shrink.
+
+      @since 0.8
+  *)
+
+  val return : 'a -> 'a t
+  (** Synonym for {!pure} *)
+
+  val make_primitive : gen : (Random.State.t -> 'a) -> shrink : ('a -> 'a Seq.t) -> 'a t
+  (** [make_primitive ~gen ~shrink] creates a generator from a function [gen] that creates
+      a random value (this function must only use the given {!Random.State.t} for randomness)
+      and a function [shrink] that, given a value [a], returns a lazy list of
+      "smaller" values (used when a test fails).
+
+      This lower-level function is meant to build generators for "primitive" types that can neither be
+      built with other primitive generators nor through composition, or to have more control on the
+      shrinking steps.
+
+      [shrink] must obey the following rules (for your own definition of "small"):
+      - [shrink a = Seq.empty] when [a] is the smallest possible value
+      - [shrink a] must return values strictly smaller than [a], ideally from smallest to largest (for
+        faster shrinking)
+      - [let rec loop a = match shrink a () with | Nil -> () | Cons (smaller_a, _) -> loop smaller_a]
+        must end for all values [a] of type ['a] (i.e. there must not be an infinite number of shrinking
+        steps).
+      
+      ⚠️ This is an unstable API as it partially exposes the implementation. In particular, the type of
+      [Random.State.t] may very well change in a future version, e.g. if QCheck switches to another
+      randomness library.
+  *)
+
+  val add_shrink_invariant : ('a -> bool) -> 'a t -> 'a t
+  (** [add_shrink_invariant f gen] returns a generator similar to [gen] except all shrinks satisfy [f].
+      This way it's easy to preserve invariants that are enforced by
+      generators, when shrinking values
+
+      @since 0.8
+
+      @deprecated is this function still useful? I feel like it is either useless (invariants
+      should already be part of the shrinking logic, not be added later) or a special,
+      incomplete case of {!Gen.t} being an Alternative (not implemented yet). For now we
+      keep it and wait for users feedback (hence deprecation to raise attention).
+  *)
+
+  (** {3 Ranges} *)
+
+  val int_bound : int -> int t
+  (** Uniform integer generator producing integers within [0..bound].
+
+      Shrinks towards [0].
+
+      @raise Invalid_argument if the argument is negative. *)
+
+  val int_range : ?origin:int -> int -> int -> int t
+  (** [int_range ?origin low high] is an uniform integer generator producing integers within [low..high] (inclusive).
+
+      Shrinks towards [origin] if specified, otherwise towards [low]
```

Comment:
do we want it to shrink towards `low` if `low < 0` and `high > 0`? or do we want to, by default, always go towards 0?

Reply:
I actually changed the functional behavior a few times on this:
* at first I aimed for the center
* then for 0
* then (current behavior) to `low`

0 is a no-go: `int_range 5 10` **must not** shrink towards 0. 0 is completely arbitrary and thus may be outside the range, so it is not a good default as it breaks invariants.

Shrinking towards the center (middle betwee low and high) felt reaaaaaaally weird when using it. I don't think it makes for a good default either.

All in all, while this **could** not be what the user expected (particularly when `low < 0`), I feel like this is the most sensible behavior.

Typically this kind of choices would bounce back to users once switching to a "real" Range feature [a la Hedgehog](https://hackage.haskell.org/package/hedgehog-1.0.5/docs/Hedgehog-Internal-Range.html#t:Range) but in the meantime, I feel like this default is the most sensible (not perfect, granted :smile: )

Reply:
I was not proposing to always converge to 0 :slightly_smiling_face: .

My proposal, that I should have made more clearly, is:

- if `low > 0` then `origin=low`
- if `high < 0` then `origin=high`
- else: 0 is in the range `[low, high]`, so `origin=0`

this way is always goes _towards_ 0, but always stays within the range.

Reply:
Ah, I understand now, thanks :+1: 

I don't have any strong feelings on that one, I'm ok with changing the behavior (here and everywhere similar) if nobody objects/has a better idea

Reply:
Done in ffe6f69be547effbfc5177f7fb69b3ec70fbcad4